### PR TITLE
Add a single public api for resolving a cloud id to a user and remote and back

### PR DESCRIPTION
**Pull request by icewind1991**
Instead of having at least 3 different implementations that might not all behave in the same way.

Additionally, having a single api point makes it easier to add more "fancy" ways to resolve the cloud id such as the various ways described in https://github.com/nextcloud/server/issues/782

### CODE REVIEW

**Comment by icewind1991** [comment]
All green, @schiessle @MorrisJobke @rullzer please review

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
Please add tests for the new classes

**Comment by icewind1991** [comment]
Rebased and added tests

**Review by MorrisJobke — APPROVED** [review]

File: apps/federatedfilesharing/lib/FederatedShareProvider.php [hunk 1]
```diff
@@ -153,17 +160,18 @@ public function create(IShare $share) {
...
+		if ($this->addressHandler->compareAddresses($cloudId->getUser(), $cloudId->getRemote(), $currentUser, $currentServer)) {
 			$message = 'Not allowed to create a federated share with the same user.';
 			$message_t = $this->l->t('Not allowed to create a federated share with the same user');
 			$this->logger->debug($message, ['app' => 'Federated File Sharing']);
 			throw new \Exception($message_t);
 		}
 
-		$share->setSharedWith($user . '@' . $remote);
+
+		$share->setSharedWith(rtrim($cloudId->getId(), '/'));
```
Comment (nickvergessen) [inline apps/federatedfilesharing/lib/FederatedShareProvider.php]: that should be done by `getId()`

File: lib/private/Federation/CloudId.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	/**
+	 * The base address of the remote server
+	 *
+	 * @return string
+	 */
+	public function getRemote() {
+		return $this->remote;
+	}
+}
```
Comment (nickvergessen) [inline lib/private/Federation/CloudId.php]: missing new line

File: lib/private/Federation/CloudId.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017, Robin Appelman <robin@icewind.nl>
+ *
```
Comment (nickvergessen) [inline lib/private/Federation/CloudId.php]: Missing ` @license GNU AGPL version 3 or any later version`

File: lib/private/Federation/CloudIdManager.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017, Robin Appelman <robin@icewind.nl>
+ *
```
Comment (nickvergessen) [inline lib/private/Federation/CloudIdManager.php]: Missing ` @license GNU AGPL version 3 or any later version`

File: lib/public/Federation/ICloudId.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	public function getUser();
+
+	/**
+	 * The base address of the remote server
+	 *
+	 * @return string
+	 * @since 12.0.0
+	 */
+	public function getRemote();
+}
```
Comment (nickvergessen) [inline lib/public/Federation/ICloudId.php]: new line

File: lib/public/Federation/ICloudId.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017, Robin Appelman <robin@icewind.nl>
+ *
```
Comment (nickvergessen) [inline lib/public/Federation/ICloudId.php]: Missing ` @license GNU AGPL version 3 or any later version`

File: lib/public/Federation/ICloudIdManager.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017, Robin Appelman <robin@icewind.nl>
+ *
```
Comment (nickvergessen) [inline lib/public/Federation/ICloudIdManager.php]: Missing ` @license GNU AGPL version 3 or any later version`

File: lib/public/Federation/ICloudIdManager.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	/**
+	 * Check if the input is a correctly formatted cloud id
+	 *
+	 * @param string $cloudId
+	 * @return bool
+	 *
+	 * @since 12.0.0
+	 */
+	public function isValidCloudId($cloudId);
+}
```
Comment (nickvergessen) [inline lib/public/Federation/ICloudIdManager.php]: new line

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Only minor things mostly, but please add it to `lib/private/AppFramework/DependencyInjection/DIContainer` so it can be auto injected in apps

**Comment by icewind1991** [comment]
@nickvergessen all fixed

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**